src/Api.js

```diff
@@ -89,13 +95,97 @@ class Api {
 	 *
-	 * @param  {string} wikiUrl URL of the wiki page that we want to analyze.
+	 * @param {HTMLElement} element
+	 * @return {int[]|boolean[]} A two-element array: 0 = token_id, 1 = user_id.
+	 */
+	getIdsFromElement( element ) {
+		const out = [ false, false ],
+			tokenMatches = element.getAttribute( 'id' ).match( /token-(\d+)/ ),
+			editorMatches = element.getAttribute( 'class' ).match( /token-editor-(\d+)/ );
+		if ( tokenMatches && tokenMatches[ 1 ] ) {
+			out[ 0 ] = parseInt( tokenMatches[ 1 ] );
+		}
+		if ( editorMatches && editorMatches[ 1 ] ) {
+			out[ 1 ] = parseInt( editorMatches[ 1 ] );
+		}
+		return out;
+	}
+
+	/**
+	 * Get the WikiWho replacement for `.mw-parser-output` HTML.
+	 * @return {string}
+	 */
+	getReplacementHtml() {
+		return this.results.extended_html;
+	}
+
+	/**
+	 * Get user and revision information for a given token.
+	 *
+	 * @param {int} tokenId
+	 * @return {{revisionId: *, score: *, userId: *, username: *, revisionTime: *}|boolean}
+	 */
+	getTokenInfo( tokenId ) {
+		let revId, revision, username, score;
+
+		// Get the token information. results.tokens structure:
+		// [ [ conflict_score, str, o_rev_id, in, out, editor/class_name, age ], ... ]
+		// e.g. Array(7) [ 0, "indicate", 769691068, [], [], "18201938", 76652371.587203 ]
+		const token = this.results.tokens[ tokenId ] || false;
+		if ( !token ) {
+			return false;
+		}
+
+		// Get revision information. results.revisions structure:
+		// { rev_id: [ timestamp, parent_rev_id, user_id, editor_name ], ... }
+		// e.g. Array(4) [ "2017-03-11T02:12:47Z", 769315355, "18201938", "Biogeographist" ]
+		revId = token[ 2 ];
+		revision = this.results.revisions[ revId ];
+		username = revision[ 3 ];
+
+		// Get the user's edit score (percentage of content edited).
+		// results.present_editors structure:
+		// [ [ username, user_id, score ], ... ]
+		for ( let i = 0; i < this.results.present_editors.length; i++ ) {
+			if ( this.results.present_editors[ i ][ 0 ] === username ) {
+				score = parseFloat( this.results.present_editors[ i ][ 2 ] ).toFixed( 1 );
+				break;
+			}
+		}
+
+		// Put it all together.
+		return {
+			username: username,
+			userId: token[ 5 ],
+			revisionId: revId,
+			revisionTime: new Date( revision[ 0 ] ),
+			score: score
+		};
+	}
+
+	/**
+	 * Get the WikiWho data for a given wiki page.
+	 *
+	 * @param {string} wikiUrl URL of the wiki page that we want to analyze.
 	 * @return {Promise}
 	 */
 	getData( wikiUrl ) {
-		return $.getJSON( this.getAjaxURL( wikiUrl ) );
+		const api = this,
+			request = $.getJSON( this.getAjaxURL( wikiUrl ) );
+		request.done( function ( result ) {
+			// Handle error response.
+			if ( !result.success && result.info && this.errorHandler ) {
+				// The API gives us an error message, but we don't use it because it's only
+				// in English.
+				return $.Deferred().reject( 'ext-whowrotethat-api-error-wikiwho-generic' );
+			}
+			// Store all results.
+			api.results = result;
```

Comment:
Why not return a construct here with the necessary items for the `.then` / `.fail`?

I guess it's a style choice and it's up to you -- we either treat `getData()` as a getter of state and always expect a promise (if it was previously rejected or resolved then attaching a .then to it runs immediately) **or** we treat it as a marker to say that the data has been fetched (which is what you're doing). Both are valid approaches, I marginally prefer the consistent promise but that's because I love promises and find the expectation consistent :stuck_out_tongue_closed_eyes: It's not crucial either way.

Whatever we do choose, we should make sure that the overlaying code has unified expectations (see my comment above about caching or flagging)

Reply:
The handler in app.js doesn't actually need the results, it just calls the other methods of Api that work with the stored results.

Reply:
> The handler in app.js doesn't actually need the results, it just calls the other methods of Api that work with the stored results.

Yeah, that's how es5-style JS singleton was done (see the activation.singleton.js file I'm overriding with this patch) -- the methods are all private, and you have an exposed method that always gives you the same instance, then calls whatever method is needed.

One of the issues with a singleton is that it's hard to do tests on it, because it's not "pure"; it's always polluted by the previous tests. I ran into this in here too which is why the tests are in this order (if you move the first test to be the second, it'll fail :\ ) I almost made the ActivationSingleton into just a function that shares an instance, but in the *actual* operation of the system it *is* supposed to be a Singleton, so I think a small payment of ordering the tests might be worth it.